src/Integration/LocalServices/ProjectSystemFilter.cs

```diff
@@ -85,13 +72,58 @@ public bool IsAccepted(DteProject dteProject)
 
         public void SetTestRegex(Regex regex)
         {
+            if (regex == null)
+            {
+                throw new ArgumentNullException(nameof(regex));
+            }
+
             this.testRegex = regex;
-            Debug.Assert(this.testRegex == null || (this.testRegex.MatchTimeout != Regex.InfiniteMatchTimeout), "Should have set non-infinite timeout");
+            Debug.Assert(this.testRegex.MatchTimeout != Regex.InfiniteMatchTimeout, "Should have set non-infinite timeout");
         }
 
         #endregion
 
         #region Helpers
+        private static bool IsTestProject(IVsHierarchy projectHierarchy, Regex testProjectNameRegex, string projectName)
+        {
+            // Otherwise, try to detect test project using known project types and/or regex match
+            if (ProjectSystemHelper.IsKnownTestProject(projectHierarchy))
+            {
+                return true;
+            }
+
+            // Heuristically exclude by project name
+            return (testProjectNameRegex != null && testProjectNameRegex.IsMatch(projectName));
+        }
+
+        private static bool IsNotSupportedProject(DteProject project)
+        {
+            var language = Language.ForProject(project);
+            return (language == null || !language.IsSupported);
+        }
+
+        private static bool IsExcludedViaProjectProperty(IVsBuildPropertyStorage propertyStorage)
+        {
+            Debug.Assert(propertyStorage != null);
+
+            // General exclusions
+            // If exclusion property is set, this takes precedence
+            bool? sonarExclude = GetPropertyBool(propertyStorage, Constants.SonarQubeExcludeBuildPropertyKey);
+            if( sonarExclude.HasValue && sonarExclude.Value)
+            {
+                return true;
+            }
+
+            // Ignore test projects
+            // If specifically marked with test project property, use that to specify if test project or not
+            bool? sonarTest = GetPropertyBool(propertyStorage, Constants.SonarQubeTestProjectBuildPropertyKey);
+            if (sonarTest.HasValue && sonarTest.Value)
```

Comment:
This is the incorrect logic change Matthew pointed out above: if _SonarQubeTestProject_ is set, then that value should be used to determine whether the project is a test project or not. See the targets file line 33-59 and 152-156: the project GUID and regular expression are only checked if an explicit value hasn't been provided.